### PR TITLE
Add util for identifying invariant axes in a WCS

### LINKED ISSUE #432: Detect Invariant Axes as part of NDCube.reproject_to
### Description

After an initial minimal implementation of `NDCube.reproject_to` has been made to work, the method needs to be able to detect invariant axes in the WCSs, i.e. axes over which no reprojection is to be performed.  This will open up a range of possibilities, such as more efficient reproject, supporting more reprojection algorithms, etc.

### PR DESCRIPTION
**Pull request by adwaitbhope**
### Description

Fixes #432.

### CODE REVIEW

File: ndcube/utils/wcs.py [hunk 1]
```diff
@@ -477,3 +478,35 @@ def compare_wcs_physical_types(source_wcs, target_wcs):
     target_wcs = get_low_level_wcs(target_wcs, 'target_wcs')
 
     return source_wcs.world_axis_physical_types == target_wcs.world_axis_physical_types
+
+
+def identify_invariant_axes(source_wcs, target_wcs, input_shape):
```
Comment (adwaitbhope) [inline ndcube/utils/wcs.py:483]: Do we need to have an `input_shape` here? Could we just use `pixel_n_dim` from the WCS and pass some static number to `np.arange()` below?

File: ndcube/utils/wcs.py [hunk 2]
```diff
@@ -477,3 +478,35 @@ def compare_wcs_physical_types(source_wcs, target_wcs):
...
+
+    input_pixel_coords = np.meshgrid(*[np.arange(n) for n in input_shape])
+
+    output_pixel_coords = pixel_to_pixel(source_wcs, target_wcs, *input_pixel_coords)
+
+    input_pixel_coords = np.around(input_pixel_coords, decimals=6)
+    output_pixel_coords = np.around(output_pixel_coords, decimals=6)
+
+    return [(input_coord == output_coord).all()
+            for input_coord, output_coord in zip(input_pixel_coords, output_pixel_coords)]
```
Comment (DanRyanIrish) [inline ndcube/utils/wcs.py:512]:
```suggestion
   return [np.allclose(input_coord, output_coord, atol=1e-6)
            for input_coord, output_coord in zip(input_pixel_coords, output_pixel_coords)]
```

File: ndcube/utils/wcs.py (hunk 1 above)
Comment (DanRyanIrish) [reply ndcube/utils/wcs.py:483]: I think we do need the `input_shape` here.

**Review by DanRyanIrish — COMMENTED** [review]
Looks good!  One suggested change at this stage.

File: ndcube/utils/wcs.py [hunk 3]
```diff
@@ -477,3 +478,32 @@ def compare_wcs_physical_types(source_wcs, target_wcs):
...
+    result: `list`
+        A list of booleans denoting whether the axis is invariant or not.
+        Follows the WCS ordering.
+    """
+
+    input_pixel_coords = np.meshgrid(*[np.arange(n) for n in input_shape])
+
+    output_pixel_coords = pixel_to_pixel(source_wcs, target_wcs, *input_pixel_coords)
+
+   return [np.allclose(input_coord, output_coord, atol=1e-6)
```
Comment (Cadair) [inline ndcube/utils/wcs.py:508]: Why atol? It might be worth exposing atol and rtol as kwargs in this function :thinking:

**Comment by Cadair** [comment]
The approach looks good. :+1:

**Comment by DanRyanIrish** [comment]
> The approach looks good. +1

Good!  Apart from your comment about exposing `atol` and `rtol`, what more needs to be done before merging this?  I *think* it's mergeable as is once that change is made?

**Comment by DanRyanIrish** [comment]
> > The approach looks good. +1
> 
> Good! Apart from your comment about exposing `atol` and `rtol`, what more needs to be done before merging this? I _think_ it's mergeable once that change is made and the tests are passing.

**Comment by adwaitbhope** [comment]
I thought the actual optimization of the `reproject_to` method would be a part of this PR itself. Should we do that here or open a separate PR?

**Comment by DanRyanIrish** [comment]
It can be.  But this seems to me to be a useful tool in its own right doesn't have to be held up waiting for the `reproject_to` optimisation.  Either way, the above changes need to be made.  So make those and we can decide then whether to merge and start a new PR or push on with this one.

**Review by DanRyanIrish — APPROVED** [review]
@Cadair, any objections to merging this?

**Pull request merged**